review/scripts/word_to_md.R

```diff
@@ -0,0 +1,443 @@
+              if (!is.null(val3) && xml_attr(third_child, "val") != val3) next
+              if (!is.null(t4)) {
+                fourth_children <- xml_children(third_child)
+                for (fourth_child in fourth_children) {
+                  if (!xml_name(fourth_child) == t4) next
+                  if (!is.null(val4) && is.na(xml_attr(fourth_child, "val"))) next
+                  if (xml_attr(fourth_child, "val") != val4) next
+                  match_found <- TRUE
+                }
+              } else match_found <- TRUE
+            }
+          } else match_found <- TRUE
+        }
+      } else match_found <- TRUE
+      
+      if (match_found) {
+        style_data <- list(
+          num_chars - nchar(xml_text(first_child)) + 1,
+          num_chars
+        )
+        xml_attr_names <- xml_attrs(first_child) %>% names()
+        if (length(xml_attr_names) & "id" %in% xml_attr_names) {
+          style_data <- c(style_data, list(xml_attrs(first_child)[["id"]]))
+        }
+        if (length(xml_attr_names) & "anchor" %in% xml_attr_names) {
+          style_data <- c(style_data, list(xml_attrs(first_child)[["anchor"]]))
+        }
+        if (as.character(i - num_blank) %in% names(smap)) {
+          prev_index <- length(smap[[as.character(i - num_blank)]])
+          new_index <- prev_index + 1
+          if (style_data[[1]] == 
+                smap[[as.character(i - num_blank)]][[prev_index]][[2]] + 1) {
+            smap[[as.character(i - num_blank)]][[prev_index]][[2]] <- style_data[[2]]
+          } else {
+            smap[[as.character(i - num_blank)]][[new_index]] <- style_data
+          }
+        } else {
+          smap[[as.character(i - num_blank)]] <- list(
+            `1` = style_data
+          )
+        }
+      }
+    }
+  }
+  
+  smap %>% map(unique)
+}
+
+maps <- list(
+  bold_map = style_map(doc, "r", "rPr", "b"),
+  ital_map = style_map(doc, "r", "rPr", "i"),
+  hypl_map = style_map(doc, "hyperlink", "r", "rPr", NULL, "rStyle", "Hyperlink"),
+  code_map = style_map(doc, "r", "rPr", "rStyle", "VerbatimChar"),
+  chyp_map = style_map(doc, "hyperlink", "r", "rPr", NULL, "rStyle", "VerbatimChar")
+)
+
+
+# Compute file breakpoints ------------------------------------------------
+
+# This will find the rows to use for each md file
+breaks <- doc_df %>%
+  filter(startsWith(text, md_dir)) %>%
+  mutate(
+    md_file = text,
+    begin   = doc_index + 1,
+    end     = lead(doc_index) - 1,
+    .keep   = "none"
+  ) %>%
+  replace_na(list(end = nrow(doc_df)))
+
+
+# Create markdown files ---------------------------------------------------
+
+# Iterate over the markdown filenames. The content for each file is transformed
+# to apply styling and hyperlinks and then written to md_out_dir
+pwalk(
+  breaks,
+  \(md_file, begin, end) {
+    # Each file has content from row number start to end
+    doc_df <- doc_df %>%
+      filter(between(doc_index, begin, end))
+
+    # Apply any bold styling
```

Comment:
Each of the 5 styling handling code blocks works very similarly. More could be done to extract this code into functions. But the prescence of edge cages such as URLs makes it a bit more complicated. So for now at least I am happy with repeating the code.